Mediacapture permissions

alvestrand opened the pull request:
This is my first attempt at incorporating mediacapture-permissions.

It has not been tidy'ed yet, since tidy did other changes to the base spec (it would be better if tidying the base spec was a separate PR).

mounirlamouri commented on index.html:
```diff
@@ -230,6 +230,27 @@
+	  all-microphones
+	</dt>
+	<dt>
+	  all-speakers
```
I wonder if we could have the "all" cases expressed in the `PermissionDescriptor` instead.

alvestrand replied:
Very possible. The descriptor is only visible as a query parameter, so omitting the deviceId should be OK. Wouldn't even change the IDL, since dictionary elements are by default optional.

mounirlamouri commented on index.html:
```diff
@@ -291,6 +312,59 @@
+	<dl class='idl' title='dictionary DevicePermissionDescriptor :
+	    PermissionDescriptor'>
+	  <dt>
+	    DOMString deviceId
```
How do you express multiple device being available/requested?

alvestrand replied:
Calling query multiple times and combining them with a permissions.AllOf()?

joeberkovitz replied:
@alvestrand Are you making an attempt to model permission to obtain device labels and metadata via device enumeration, so that the user can make meaningful choices of device?

alvestrand replied:
When people have access to the list of devices via enumerateDevices(), they should be able to use that information together with query() to get information about permission to those devices, yes. @joeberkovitz was that what you were asking, or did I misunderstand your question?

stefhak commented on index.html:
```diff
@@ -291,6 +312,59 @@
+	  status.
+	</p>
+	<p>
+	  A successful call to the <code>GetUserMedia</code> function of
```
lowercase "G"?
Strictly speaking I think there are several ways to call (invoke?) getUserMedia, but perhaps we do not need to be more specific that this.

stefhak commented on index.html:
```diff
@@ -291,6 +312,59 @@
+	</p>
+	<p>These three permissions have an associated PermissionDescriptor:
+	</p>
+	<dl class='idl' title='dictionary DevicePermissionDescriptor :
```
Should this be _Media_DevicePermissionDescriptor (or do we believe all future devices can use this model)?

stefhak commented on index.html:
```diff
@@ -291,6 +312,59 @@
+	  and <code><dfn id='idl-def-PermissionName.speaker'>speaker</dfn>
+	  </code>
+	  permissions are associated with permission to use media
+	  devices as specified in [[mediacapture-main]].
```
Aren't permissions for speakers described in mediacapture-output?

stefhak commented:
I think a media device example should be added. Otherwise, apart from the comments above, this LGTM.

stefhak commented on index.html:
```diff
@@ -291,6 +312,59 @@
+	  and output devices.
+	</p>
+	<p>
+	  The permission storage identifier for permissions in this
```
Trying to parse this. In https://github.com/w3c/mediacapture-main/issues/267 it seems we're zooming in to keying by origin and top-level origin.
Here we we say "origin and embedder's origin". Assuming we have more than one level of iFrame's, this would mean a difference between the two(?).

mounirlamouri replied:
+1, top level sounds better.

alvestrand commented:
PTAL

mounirlamouri commented on index.html:
```diff
@@ -291,6 +303,63 @@
+	  at the moment or are attached at a later time. When such a
+	  permission has been granted, queries for permission to any
+	  specific device MUST return "granted" as long as the
+	  deviceId is the ID of a connected device of the given class.
```
You don't explain what to do in the case of `deviceId` being something else. Is that on purpose. I particularly wonder what we should do if it is the empty string.

alvestrand replied:
Hm... for the case where it's not '*' and it's not a legal deviceId?

No, that's not specified - I thought the API would have a general text saying "if the parameters are illegal, the result is an error" - but there doesn't seem to be any such language; when retrieving a permission object that doesn't exist, the result is implementation-defined, but always returning "prompt" is one suggested resolution.

I'd suggest an error for illegal descriptors. WDYT?

adam-be commented on index.html:
```diff
@@ -291,6 +303,63 @@
+	  be granted.
+	</p>
+	<p>
+	  Closing a MediaStreamTrack MAY cause permission to be
```
s/Closing/Stopping/ may be a better match to the language in the Media Capture and Streams spec.

mounirlamouri replied:
What about this comment?

alvestrand replied:
Done.

martinthomson commented on index.html:
```diff
@@ -291,6 +303,63 @@
+	<p>
+	  If the deviceId is the special string "*", it queries for
+	  access to all devices of that class, whether they exist
+	  at the moment or are attached at a later time. When such a
```
Is the difference between an absent deviceId and a deviceId of "*" that absence is "devices of this class that are here now" and "*" means "devices of this class, no matter their current presence"

alvestrand replied:
The first one means that getusermedia(...) (no deviceid constraint) will succeed, but you don't know which device you're getting - there might be other cameras you don't have permission to access.

The second means that getusermedia({video: {deviceId: xxx}}) will succeed as long as xxx is a legal deviceid.

(it can fail for other reasons than permission, of course. I'm simplifying).

But as I said, I'm not sure the extra complexity is really worth it.

martinthomson replied:
I think that this way of expressing the "any" vs. "all" distinction is as clear and transparent as mud.

I'm still uncertain about the benefits, but if we do this, we need a more explicit way of making the distinction between specific device, any device and all devices.

alvestrand replied:
Please suggest text... for instance, we could use the strings "some device" for <empty> and "all devices" for "*" - this revisits the question of what values of a device ID can be used safely for something other than a device id - or we could add another attribute.
My first proposal (separate "all-" permissions) was a bit easier to explain, but did not support the "any" case.

mounirlamouri replied:
"all" by default concerns me because it could create issues for browsers that don't do all by default with getUserMedia() calls. However, with permissions.request() we could make this behaviour clearer.

martinthomson@, if you are fine with the possible compat issues that Firefox could run into, I'm fine with removed the distinction between "any" and "all".

stefhak commented on index.html:
```diff
@@ -291,6 +303,63 @@
+	<p>
+	  The permission storage identifier for permissions in this
+	  section MUST include the top level context's origin and the embedding
+	  status.
```
(just to make sure:) Is there a definition of "top level context's origin" somewhere, or is it something that can't be misunderstood?
In the spec there is also a sentence in section 7 that ends "...MAY contain other information like the embedding status or the embedder's origin." Should we add top level context's origin here?

alvestrand replied:
I think we're talking about https://www.w3.org/html/wg/drafts/html/master/browsers.html#top-level-browsing-context - every time I try to follow this chain of references, my head hurts, but I can't quite see how it could be anything else.

alvestrand commented:
Two changes after discussions:
- Removed the "any" specifier. It was too confusing to reason about what it really meant. The "all" specifier is reasonably easy to reason about, so I described it a little more.
- Removed the scope specification. This is ongoing discussion, and we don't want to add language that makes resolution of ongoing discussions more difficult.

mounirlamouri commented on index.html:
```diff
@@ -291,6 +303,60 @@
+	<p>
+	  If the descriptor does not have a deviceId, its semantic is that
+	  it queries for access to all devices of that class. Thus, if a query
+          for the "camera" permission with no deviceId returns "granted", the
```
nit: alignment issue.

mounirlamouri commented:
lgtm with the two comments resolved. Also, could you rebase the commits to one?

alvestrand commented:
One commit, and tidy done. My version of tidy prefers </li> on its own line, it seems. Apologies for the diff noise.

mounirlamouri commented:
tidy shouldn't do that. You might be using another version of `tidy-html5` or you didn't used the `tidyconfig.txt` file. Note that I saw that `contributing.md` instructions wasn't matching the file name so it could be that.

I know it's painful but I would appreciate if you could clean up the changes.

mounirlamouri commented:
Thank you for your patience! \o/

alvestrand commented:
Arrggghhhh - with all the merging, rebasing, tidying and untidying, the last set of changes (removing the * and the special language for top-level origin) have disappeared from the PR :-( - I'll make a new one.
